### PR TITLE
JavaScript: Avoid duplicate source/sink nodes due to flow labels.

### PR DESCRIPTION
**Pull request by max-schaefer**
Flow labels form part of the identity of a `PathNode`, leading to duplicate sources and/or sinks in queries that use them (though some of our tools have settings to collapse them).

This PR proposes to fix this by introducing synthetic source and sink nodes that do not incorporate information about flow labels, and have the real source/sink as their successor/predecessor. The real source and sink are hidden in the path viewer to avoid having confusing edges between identical-looking nodes.

A [preliminary dist-compare](https://git.semmle.com/max/dist-compare-reports/blob/master/js/avoid-duplicate-source-and-sink-nodes/report.md) on nightly.slugs shows no effects on performance or results, but I'm running a full dist-compare just in case.

Another thing I'd like to point out is that this PR turns paths of length zero into paths of length one: if a source node is also a sink node, we still introduce both a synthetic source and a synthetic sink with an edge between them (skipping over the original source-cum-sink node). This is less nice than before, but I don't think it happens that much in practice, and avoiding it is a bit fiddly. I'm happy to be convinced otherwise, though.

Commit-by-commit review is strongly encouraged.

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 1]
```diff
@@ -994,10 +994,28 @@ private predicate isLive() { exists(DataFlow::Configuration cfg | isSource(_, cf
...
+  MkSourceNode(DataFlow::Node nd, DataFlow::Configuration cfg) { mkSourceNode(nd, cfg, _) }
+  or
   MkMidNode(DataFlow::Node nd, DataFlow::Configuration cfg, PathSummary summary) {
     isLive() and
     onPath(nd, cfg, summary)
   }
+  or
+  MkSinkNode(DataFlow::Node nd, DataFlow::Configuration cfg) { mkSinkNode(nd, cfg, _) }
+
+private predicate mkSourceNode(DataFlow::Node nd, DataFlow::Configuration cfg, PathSummary summary) {
```
[the remark this answers is not in the document]
Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: It's also badly named; will fix.

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 2]
```diff
@@ -1099,31 +1139,27 @@ class MidPathNode extends PathNode, MkMidNode {
     // Skip to the top of big left-leaning string concatenation trees.
     nd = any(AddExpr add).flow() and
     nd = any(AddExpr add).getAnOperand().flow()
+    or
+    // Skip mid node immediately following a source node
+    exists(MkSourceNode(nd, cfg))
```
[the remark this answers is not in the document]
Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: That is a very good point. I was going to hide the edges (not the nodes) in the follow-up PR to do with path exploration; I'll just add that commit to this PR instead.

File: javascript/ql/src/semmle/javascript/dataflow/Configuration.qll [hunk 3]
```diff
@@ -1083,17 +1064,51 @@ private PathNode getASuccessor(PathNode pnd) {
 }
 
 private PathNode getASuccessorIfHidden(PathNode nd) {
-  nd.isHidden() and
+  nd.(MidPathNode).isHidden() and
   result = getASuccessor(nd)
 }
 
+/**
```
Comment (esbena) [inline javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: Is the diff bad, or did you include two qldoc starters here?
```
/**
/**
```

Comment (max-schaefer) [reply javascript/ql/src/semmle/javascript/dataflow/Configuration.qll]: The latter; will fix.

**Comment by asger-semmle** [comment]
The new commits LGTM. It's a bit unfortunate that it wasn't part of the evaluation, though. Could you make sure it's exercised in at least one evaluation at some point (possibly together with the next PR as per your original plan).

**Comment by max-schaefer** [comment]
I'll redo the whole evaluation. I'd like to get as much validation as possible, considering how many mistakes I made trying to get the right edges suppressed in that last commit :disappointed:

**Comment by max-schaefer** [comment]
A [re-run](https://git.semmle.com/max/dist-compare-reports/blob/master/js/avoid-duplicate-source-and-sink-nodes/report.md) (internal link) on nightly.slugs shows no impact on results or performance.

**Comment by max-schaefer** [comment]
Same story for default.slugs, so let's give this a try.

**Pull request merged**